cpu/stm32_common/periph/flashpage.c

```diff
@@ -89,5 +91,86 @@ void flashpage_write(int page, void *data)
         while (RCC->CR & RCC_CR_HSIRDY) {}
     }
 }
+#elif defined(CPU_FAM_STM32L0) || defined(CPU_FAM_STM32L1)
+/* Flash program erase key1 */
+#define PEKEY1                    (0x89ABCDEF)
+
+/* Flash program erase key: used with FLASH_PEKEY2 to unlock the write
+ * access to the FLASH_PECR register and data EEPROM
+ */
+#define PEKEY2                    (0x02030405)
+
+/* Flash program memory key1 */
+#define PRGKEY1                   (0x8C9DAEBF)
+
+/* Flash program memory key2: used with FLASH_PRGKEY2 to unlock
+ * the program memory
+ */
+#define PRGKEY2                   (0x13141516)
+
+void flashpage_write(int page, void *data)
+{
+    assert(page < FLASHPAGE_NUMOF);
+
+    /* Using 32bit addresses since we are attempting fast-word erasing/writing*/
+    uint32_t *page_addr = flashpage_addr(page);
+    uint32_t *data_addr = (uint32_t*)data;
+
+    DEBUG("[flashpage] unlocking the flash module\n");
+    /* Unlocking the Data memory and FLASH_PECR register access*/
+    if(FLASH->PECR & FLASH_PECR_PRGLOCK)
+    {
+        /* Unlock for erase */
+        FLASH->PEKEYR = PEKEY1;
+        FLASH->PEKEYR = PEKEY2;
+
+        /* Unlock flash registers */
+        FLASH->PRGKEYR = PRGKEY1;
+        FLASH->PRGKEYR = PRGKEY2;
+    }
+
+    /* ERASE sequence */
+    /* make sure no flash operation is ongoing */
+    DEBUG("[flashpage] erase: waiting for any operation to finish\n");
+    while (FLASH->SR & FLASH_SR_BSY) {}
+
+    DEBUG("[flashpage] erase: setting the erase and program bits\n");
+    /* Set the ERASE bit */
+    FLASH->PECR |= FLASH_PECR_ERASE;
+    /* Set PROG bit */
+    FLASH->PECR |= FLASH_PECR_PROG;
+
+    DEBUG("address to erase: %p\n", flashpage_addr(page));
+    DEBUG("[flashpage] erase: trigger the page erase\n");
+    /* Write 00000000h to the first word of the program page to erase */
+    *page_addr = 0x00000000;
+    /* Wait for it to be finished */
+    DEBUG("[flashpage] erase: wait as long as device is busy\n");
+    while (FLASH->SR & FLASH_SR_BSY) {}
+
+    /* If the erase operation is completed, disable the ERASE and PROG bits */
+    DEBUG("[flashpage] erase: resetting the page erase and prog bit\n");
+    FLASH->PECR &= (uint32_t)(~FLASH_PECR_PROG);
```

Comment:
hm, are you sure about this one? You also disable it in line 168, and shouldn't it only be done there, after you have (potentially) written to flash?

Reply:
Oops, a leftover of a test, will remove.

Reply:
It is working like that, since the program bit apparently only affects when erasing the flash.